server/kafka/partitioner.go

```diff
@@ -0,0 +1,72 @@
+/*
+ * Copyright 2020 The NATS Authors
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+package kafka
+
+import (
+	"github.com/Shopify/sarama"
+	"sync"
+)
+
+// A balancer (or partitioner in Sarama terms) is in-charge of spreading messages across available partitions of a topic.
+// Sarama provides a hash based partitioner which is default in the producer. This implementation distributes messages
+// based on number of bytes each partition has received.
+type leastBytesPartitioner struct {
+	mutex        sync.Mutex
+	byteCounters map[int]uint64
+}
+
+func NewLeastBytesPartitioner(topic string) sarama.Partitioner {
+	lbp := new(leastBytesPartitioner)
+	lbp.byteCounters = make(map[int]uint64)
+	return lbp
+}
+
+func (lbp *leastBytesPartitioner) RequiresConsistency() bool {
+	return false
+}
+
+func (lbp *leastBytesPartitioner) Partition(message *sarama.ProducerMessage, numPartitions int32) (int32, error) {
+	numPartitionsInt := int(numPartitions)
+
+	lbp.mutex.Lock()
+	defer lbp.mutex.Unlock()
+
+	// if partition count has reduced, remove the old entries
+	for i := len(lbp.byteCounters) - 1; i >= numPartitionsInt; i-- {
+		delete(lbp.byteCounters, i)
+	}
+
+	// find the entry in the byteCounters with min bytes
+	minIndex := findPartitionWithMinBytes(lbp.byteCounters)
+	lbp.byteCounters[minIndex] += uint64(message.Key.Length()) + uint64(message.Value.Length())
+
+	return int32(minIndex), nil
+}
+
+func findPartitionWithMinBytes(counters map[int]uint64) int {
+	var minPart int
+	var minBytes uint64
```

Comment:
Can we add a test to this function? Maybe call it `TestFindPartitionWithMinBytes` and put it in `partitioner_test.go`.

Reply:
Done.